### PR TITLE
[CALCITE-6020] SqlToRelConverter should not replace windowed SUM with…

### PR DESCRIPTION
**Pull request by normanj-bitquill**
… equivalent expression using SUM0

* Add a new configuration option to RelBuilder.Config
  * Whether to change SUM to SUM0 in a window
* Call over to RelBuilder to determine the operator to use when in a window but not a histogram

### CODE REVIEW

**Comment by mihaibudiu** [comment]
This is supposedly fixing a bug. No test for that bug?

**Comment by normanj-bitquill** [comment]
I've reworked this PR. Now the conversion of SUM to SUM0 happens in a new RelRule.

The advantage is that a user can choose whether to include the new RelRule or not.
The downside is that a user could lose the SUM to SUM0 conversion on upgrade depending on how they build up their list of RelRules to use.

File: core/src/test/resources/org/apache/calcite/test/SqlToRelConverterTest.xml [hunk 1]
```diff
@@ -5790,7 +5790,7 @@ window w1 as (partition by job order by hiredate rows 2 preceding)]]>
     </Resource>
     <Resource name="plan">
       <![CDATA[
-LogicalProject(EXPR$0=[CASE(>(COUNT($5) OVER (PARTITION BY $2 ORDER BY $4 ROWS 2 PRECEDING), 0), $SUM0($5) OVER (PARTITION BY $2 ORDER BY $4 ROWS 2 PRECEDING), null:INTEGER)], EXPR$1=[/(CASE(>(COUNT(CAST($5):REAL NOT NULL) OVER (PARTITION BY $2 ORDER BY $4 ROWS 2 PRECEDING), 0), $SUM0(CAST($5):REAL NOT NULL) OVER (PARTITION BY $2 ORDER BY $4 ROWS 2 PRECEDING), null:REAL), COUNT(CAST($5):REAL NOT NULL) OVER (PARTITION BY $2 ORDER BY $4 ROWS 2 PRECEDING))])
+LogicalProject(EXPR$0=[CASE(>(COUNT($5) OVER (PARTITION BY $2 ORDER BY $4 ROWS 2 PRECEDING), 0), SUM($5) OVER (PARTITION BY $2 ORDER BY $4 ROWS 2 PRECEDING), null:INTEGER)], EXPR$1=[/(CASE(>(COUNT(CAST($5):REAL NOT NULL) OVER (PARTITION BY $2 ORDER BY $4 ROWS 2 PRECEDING), 0), SUM(CAST($5):REAL NOT NULL) OVER (PARTITION BY $2 ORDER BY $4 ROWS 2 PRECEDING), null:REAL), COUNT(CAST($5):REAL NOT NULL) OVER (PARTITION BY $2 ORDER BY $4 ROWS 2 PRECEDING))])
```
[the remark this answers is not in the document]
Comment (mihaibudiu) [reply core/src/test/resources/org/apache/calcite/test/SqlToRelConverterTest.xml:5793]: Or maybe this is actually addressing the actual issue: you are saying that theses were incorrect to start with?

**Review by mihaibudiu — APPROVED** [review]

File: core/src/main/java/org/apache/calcite/rel/rules/ProjectOverSumToSum0Rule.java [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+import org.apache.calcite.sql.fun.SqlStdOperatorTable;
+import org.apache.calcite.tools.RelBuilder;
+
+import org.immutables.value.Value;
+
+import java.util.List;
+
+/**
+ * Planner rule that converts SUM to SUM0 when it is used in an OVER clause inside
+ * the project list.
```
[the remark this answers is not in the document]
Comment (normanj-bitquill) [reply core/src/main/java/org/apache/calcite/rel/rules/ProjectOverSumToSum0Rule.java]: I expanded on the Javadoc here to explain why this rule should be used. Let me know if more explanation is needed.

File: core/src/main/java/org/apache/calcite/rex/RexShuttle.java [hunk 3]
```diff
@@ -41,16 +43,17 @@ public class RexShuttle implements RexVisitor<RexNode> {
   @Override public RexNode visitOver(RexOver over) {
     boolean[] update = {false};
     List<RexNode> clonedOperands = visitList(over.operands, update);
+    SqlAggFunction op = visitOverAggFunction(over.getAggOperator());
```
[the remark this answers is not in the document]
Comment (normanj-bitquill) [reply core/src/main/java/org/apache/calcite/rex/RexShuttle.java]: Updated.

File: core/src/test/resources/org/apache/calcite/test/SqlToRelConverterTest.xml (hunk 1 above)
[the remark this answers is not in the document]
Comment (normanj-bitquill) [reply core/src/test/resources/org/apache/calcite/test/SqlToRelConverterTest.xml:5793]: This is sort of a regression. These tests only convert the SQL to Rel, but do not go through planning the query. As a result, they do not run the RelRules, so `SUM` is not converted to `SUM0`. There are other tests (in `RelOptRulesTest`) that also go through planning and verify that `SUM` becomes `SUM0`.

**Comment by normanj-bitquill** [comment]
The commits have been squashed.

**Pull request merged**